engine/table/src/main/java/io/deephaven/engine/table/impl/remote/ConstructSnapshot.java

```diff
@@ -1613,57 +1749,87 @@ private static <T> Object getSnapshotData(
-                final ChunkType chunkType = columnSource.getChunkType();
-
-                // create a new chunk
-                WritableChunk<Values> currentChunk = chunkType.makeWritableChunk(chunkSize);
-
-                if (usePrev) {
-                    columnSource.fillPrevChunk(context, currentChunk, reducedRowSet);
-                } else {
-                    columnSource.fillChunk(context, currentChunk, reducedRowSet);
+                if (reducedRowSet.intSize() != chunkSize) {
+                    failIfConcurrentAttemptInconsistent();
+                    throw new SnapshotUnsuccessfulException(String.format(
+                            "Rowset did not provide requested number of rows in an otherwise consistent state, " +
+                                    "requested = %d, actual = %d",
+                            chunkSize, reducedRowSet.size()));
                 }
-
-                // add the chunk to the current list
-                result.add(currentChunk);
-
-                // increment the offset for the next chunk (using the actual values written)
-                offset += currentChunk.size();
+                // Populate the snapshot data for each column for the current chunk of rows
+                for (int colRank = 0; colRank < numCols; ++colRank) {
+                    final int colIdx = columnIndices.get(colRank);
+                    final ColumnSource<?> columnSource = columnSources.get(colRank);
+                    final ColumnSource.FillContext fillContext = fillContexts[colRank];
+                    // Populate snapshot data as a list of chunks
+                    if (snapshot.addColumnData[colIdx].data == null) {
+                        snapshot.addColumnData[colIdx].data = new ArrayList<>();
+                    }
+                    snapshot.addColumnData[colIdx].data.add(
+                            getColumnSnapshotChunk(columnSource, fillContext, reducedRowSet, chunkSize, usePrev));
+                }
+                // increment offset for the next chunk (assuming the number of rows written for all columns is equal)
+                offset += chunkSize;
 
                 // recompute the size of the next chunk
-                if (size - offset > maxChunkSize) {
+                if (numRows - offset > maxChunkSize) {
                     chunkSize = maxChunkSize;
                 } else {
-                    chunkSize = (int) (size - offset);
+                    chunkSize = (int) (numRows - offset);
                 }
-
                 if (sharedContext != null) {
                     // a shared context is good for only one chunk of rows
                     sharedContext.reset();
                 }
             }
         }
-        return result;
+    }
+
+    /**
+     * Get a chunk of data for provided column source and row set.
+     */
+    private static <T> Chunk<Values> getColumnSnapshotChunk(
```

Comment:
I think this method is a little silly. Firstly, you cannot have a 0-length chunk in the only caller. Secondly, it's almost thrice as long as if you just in-lined it once you remove that unnecessary bit.